### PR TITLE
Specialization to save additions in Fp2 for negative non-residue

### PR DESCRIPTION
**Pull request by ValarDragon**
In particular, this adds add_and_mul_fp_by_nonresidue, sub_and_mul_fp_by_nonresidue
and utilizes this in fp2 multiplication and squaring, saving additions in each.

This is because when the non-residue is negative, there is an extra operation for negating it. However every usage is of the form `x + mul_by_non_residue(y)` or `x - mul_by_non_residue(y)`, so that final negation can be absorbed.

I've benchmarked this on bls12_377, and this is an 8% savings to square and a 4% savings to mul_assign.

```
 name                             master.txt ns/iter  fp2_optimization.txt ns/iter  diff ns/iter  diff %  speedup
 bls12_377::fq2::add_assign       15                       15                                   0   0.00%   x 1.00
 bls12_377::fq2::deser            155                      176                                 21  13.55%   x 0.88
 bls12_377::fq2::deser_unchecked  155                      176                                 21  13.55%   x 0.88
 bls12_377::fq2::double           13                       12                                  -1  -7.69%   x 1.08
 bls12_377::fq2::inverse          11,891                   11,650                            -241  -2.03%   x 1.02
 bls12_377::fq2::mul_assign       232                      223                                 -9  -3.88%   x 1.04
 bls12_377::fq2::negate           24                       24                                   0   0.00%   x 1.00
 bls12_377::fq2::ser              90                       91                                   1   1.11%   x 0.99
 bls12_377::fq2::ser_unchecked    90                       94                                   4   4.44%   x 0.96
 bls12_377::fq2::sqrt             149,598                  150,679                          1,081   0.72%   x 0.99
 bls12_377::fq2::square           222                      203                                -19  -8.56%   x 1.09
 bls12_377::fq2::sub_assign       16                       16                                   0   0.00%   x 1.00
```
(I believe the serialization and sqrt speed differences are noise, since they were untouched. However it is slightly suspicious to claim this, since deser took a significant overhead. But I don't see why else deser_unchecked would have slowed down as it doesn't touch any of the involved methods)

Benchmarks for g2 are currently running, thought I'd open the PR in case anyone wanted to start reviewing it sooner.

/cc @jon-chuang

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests - should be covered by existing tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by ValarDragon** [comment]
g2 performance: 
```
 name                             master.txt ns/iter  fp2_optimization.txt ns/iter  diff ns/iter  diff %  speedup
 bls12_377::g2::add_assign        4,177               4,012                            -165  -3.95%   x 1.04
 bls12_377::g2::add_assign_mixed  2,935               2,809                            -126  -4.29%   x 1.04
 bls12_377::g2::deser             901,001             879,512                       -21,489  -2.39%   x 1.02
 bls12_377::g2::deser_unchecked   315                 315                                 0   0.00%   x 1.00
 bls12_377::g2::double            1,949               1,891                             -58  -2.98%   x 1.03
```

bls12_381 fp2 numbers (7% multiplication speedup, 9% square speedup. I believe double and negate times are noise):
```
 bls12_381::fq2::add_assign       16                                 15                                            -1  -6.25%   x 1.07
 bls12_381::fq2::deser            153                                151                                           -2  -1.31%   x 1.01
 bls12_381::fq2::deser_unchecked  154                                151                                           -3  -1.95%   x 1.02
 bls12_381::fq2::double           12                                 13                                             1   8.33%   x 0.92
 bls12_381::fq2::inverse          11,797                             11,656                                      -141  -1.20%   x 1.01
 bls12_381::fq2::mul_assign       216                                199                                          -17  -7.87%   x 1.09
 bls12_381::fq2::negate           23                                 24                                             1   4.35%   x 0.96
 bls12_381::fq2::ser              90                                 90                                             0   0.00%   x 1.00
 bls12_381::fq2::ser_unchecked    92                                 90                                            -2  -2.17%   x 1.02
 bls12_381::fq2::sqrt             110,087                            109,968                                     -119  -0.11%   x 1.00
 bls12_381::fq2::square           172                                156                                          -16  -9.30%   x 1.10
 bls12_381::fq2::sub_assign       16                                 16                                             0   0.00%   x 1.00
```

altbn254 fp2 numbers (mul assign +5%, square down 10%)
```
 bn254::fq2::add_assign       11                             10                                        -1   -9.09%   x 1.10
 bn254::fq2::deser            71                             72                                         1    1.41%   x 0.99
 bn254::fq2::deser_unchecked  71                             73                                         2    2.82%   x 0.97
 bn254::fq2::double           13                             13                                         0    0.00%   x 1.00
 bn254::fq2::inverse          5,646                          5,878                                    232    4.11%   x 0.96
 bn254::fq2::mul_assign       106                            111                                        5    4.72%   x 0.95
 bn254::fq2::negate           15                             17                                         2   13.33%   x 0.88
 bn254::fq2::ser              39                             62                                        23   58.97%   x 0.63
 bn254::fq2::ser_unchecked    39                             62                                        23   58.97%   x 0.63
 bn254::fq2::sqrt             36,115                         36,100                                   -15   -0.04%   x 1.00
 bn254::fq2::square           94                             84                                       -10  -10.64%   x 1.12
 bn254::fq2::sub_assign       11                             11                                         0    0.00%   x 1.00
```
Not sure why mul assign got worse for bn254 fp2, but square did get better =/

**Comment by ValarDragon** [comment]
it does make sense that the speedup is more pronounced for bls12_381 and bn254 (since their non-residue is `-1` whereas bls12_377 is `-5`), but I don't understand why bn254's mul got worse.

**Comment by ValarDragon** [comment]
Rebenchmarked the mul_assign for bn254 and got that its breakeven at 109ns. I benchmarked the g2 ops to get a better of sense how this affects speed in a greater scope:

```
 name                         control_bn254_g2.txt ns/iter  variable_bn254_g2.txt ns/iter  diff ns/iter  diff %  speedup
 bn254::g2::add_assign        2,029                         1,989                                   -40  -1.97%   x 1.02
 bn254::g2::add_assign_mixed  1,443                         1,411                                   -32  -2.22%   x 1.02
 bn254::g2::deser             412,998                       406,374                              -6,624  -1.60%   x 1.02
 bn254::g2::deser_unchecked   145                           150                                       5   3.45%   x 0.97
 bn254::g2::double            957                           933                                     -24  -2.51%   x 1.03
```

It speeds up the relevant g2 operations by 2-2.5%, so I feel like its fine to conclude that this is real optimization? Thoughts?

**Comment by Pratyush** [comment]
Are the results within the realm of noise? like, do we consistently get those speedups, or? Another way to benchmark the effect would be to benchmark the cost of fq12 ops

File: ff/src/fields/models/fp2.rs
Comment (Pratyush) [inline ff/src/fields/models/fp2.rs:15]:
```suggestion
    /// Return `fe * Self::NONRESIDUE`.
```

File: ff/src/fields/models/fp2.rs
Comment (Pratyush) [inline ff/src/fields/models/fp2.rs:21]:
```suggestion
    /// A specializable method for computing `x + mul_base_field_by_nonresidue(y)`
```

File: ff/src/fields/models/fp2.rs
Comment (Pratyush) [inline ff/src/fields/models/fp2.rs:29]:
```suggestion
    /// A specializable method for computing `x - mul_base_field_by_nonresidue(y)`
```

**Comment by ValarDragon** [comment]
I re-ran each of the fp2 benchmarks and got pretty similar results (numbers within +-3 ns), with exception of the bn254 mul_assign. (And serialization always having inconsistent times)

I'll benchmark the fp12 arithmetic so we get more results

**Comment by ValarDragon** [comment]
fq12 benchmarks
```
 bls12_377::fq12::mul_assign       6,127                     6,007                              -120  -1.96%   x 1.02
 bls12_377::fq12::square           4,280                     4,194                               -86  -2.01%   x 1.02
 bls12_381::fq12::mul_assign       5,452                     5,359                               -93  -1.71%   x 1.02
 bls12_381::fq12::square           3,804                     3,728                               -76  -2.00%   x 1.02
 bn254::fq12::mul_assign           3,748                     3,759                                11   0.29%   x 1.00
 bn254::fq12::square               2,815                     2,725                               -90  -3.20%   x 1.03
```
We're seeing everything speedup except bn254::fq12:;mul_assign, which increased by .3%.

**Comment by Pratyush** [comment]
btw, is something similar possible for cubic extensions?

**Comment by ValarDragon** [comment]
Looking at the extremely well documented code in constantine, we can actually further customize the squaring routine to save 2 more additions and 1 temporary for bn254 and bls12-381 (when `nonresidue=-1`). I'll implement this and see if it speeds it up. https://github.com/mratsim/constantine/blob/2c5e12d5f893c6af9b97fec17d572268de3bd899/constantine/tower_field_extensions/quadratic_extensions.nim#L17

Re: Cubic extensions, I didn't notice any of the fields using a negative cubic extension, so I don't see a way at the moment to improve the mul by non-residue portion of it.

**Comment by ValarDragon** [comment]
Latest commit further speeds up squaring for curves with non-residue=-1, new speeds are:
```
 bls12_381::fq2::add_assign       15                           15                                       0    0.00%   x 1.00
 bls12_381::fq2::double           12                           13                                       1    8.33%   x 0.92
 bls12_381::fq2::inverse          11,814                       11,923                                 109    0.92%   x 0.99
 bls12_381::fq2::mul_assign       210                          201                                     -9   -4.29%   x 1.04
 bls12_381::fq2::negate           23                           23                                       0    0.00%   x 1.00
 bls12_381::fq2::sqrt             110,139                      111,321                              1,182    1.07%   x 0.99
 bls12_381::fq2::square           171                          141                                    -30  -17.54%   x 1.21
 bls12_381::fq2::sub_assign       16                           16                                       0    0.00%   x 1.00
 bn254::fq2::add_assign       10                        10                                   0    0.00%   x 1.00
 bn254::fq2::double           13                        13                                   0    0.00%   x 1.00
 bn254::fq2::inverse          5,630                     5,645                               15    0.27%   x 1.00
 bn254::fq2::mul_assign       106                       110                                  4    3.77%   x 0.96
 bn254::fq2::negate           15                        16                                   1    6.67%   x 0.94
 bn254::fq2::sqrt             35,142                    35,149                               7    0.02%   x 1.00
 bn254::fq2::square           92                        73                                 -19  -20.65%   x 1.26
 bn254::fq2::sub_assign       12                        11                                  -1   -8.33%   x 1.09
 bls12_377::fq2::add_assign       15                       15                                       0   0.00%   x 1.00
 bls12_377::fq2::double           13                       12                                      -1  -7.69%   x 1.08
 bls12_377::fq2::inverse          11,891                   11,671                                -220  -1.85%   x 1.02
 bls12_377::fq2::mul_assign       232                      223                                     -9  -3.88%   x 1.04
 bls12_377::fq2::negate           24                       24                                       0   0.00%   x 1.00
 bls12_377::fq2::sqrt             149,598                  150,469                                871   0.58%   x 0.99
 bls12_377::fq2::square           222                      204                                    -18  -8.11%   x 1.09
 bls12_377::fq2::sub_assign       16                       16                                       0   0.00%   x 1.00
```

So square for bls12_381 and bn254 got another ~8% faster, and speed is the same (as expected) for bls12_377

**Comment by ValarDragon** [comment]
I personally view this as ready to merge, seems to not be noise to me given that its improved other more complex benchmarks (g2, fp12), and is quite significant for square

File: ff/src/fields/models/quadratic_extension.rs
Comment (Pratyush) [inline ff/src/fields/models/quadratic_extension.rs:258]: I think it's worthwhile here to add the complex multiplication formula, and then show how it's optimized in the `P::NONRESIDUE == -1` case.

**Comment by Pratyush** [comment]
LGTM once the complex multiplication comment is in

**Comment by Pratyush** [comment]
Btw, does this make sense for the higher extensions also (Fp12/Fp62over3)?

File: ff/src/fields/models/quadratic_extension.rs
Comment (ValarDragon) [reply ff/src/fields/models/quadratic_extension.rs]: Added, this is a great suggestion! Also it pointed out another field op we can save in squaring for bls12_377

**Comment by ValarDragon** [comment]
I don't at the moment know of a similar trick for Fp12/Fp62over3, but its pretty plausible they exist

**Comment by ValarDragon** [comment]
Latest commit does the additional optimization mentioned in issue comment https://github.com/arkworks-rs/algebra/pull/207#discussion_r571329087.

This further reduces the time for a square in bls12_377 by an additional 8%. (15.77% total). I confirmed that that change did not affect other fp2 fields. (e.g. ones with -1 or without the optimization added)

**Comment by jon-chuang** [comment]
I'm not convinced these extra methods with verbose names is necessary to achieve what you are aiming for

**Comment by ValarDragon** [comment]
Not really sure how to, without a #206 style solution. (In which case we could remove `mul_by_nonresidue` as well.

**Comment by Pratyush** [comment]
@valardragon, if we had #206, how would that help us for this?

**Comment by ValarDragon** [comment]
Its not entirely clear to me at the moment. If 206 had support for negative scalars, we could maybe remove custom methods on fp2 for cases where the nonresidue is small. However some fields have large non-residues, and we'd still need this to support ideal addition chains there.

**Comment by Pratyush** [comment]
Negative scalars can be supported via just one negation, no? And AFAIK it should be easy to compute addition chains for all small scalars (let's say 17, because that's the largest one we have)

I'm wondering whether we can replace the NONRESIDUE by `i64` in the relevant places, and use that for the towers

**Comment by ValarDragon** [comment]
Yeah thats true for negative scalars. I don't really see a need to block this PR on having #206 API figured out, unless someones planning on doing it in the next couple of days. (It'd be a breaking change anyway, since we'd have to remove the existing `mul_by_nonresidue` methods from the `Fp2Parameters` trait, and instead only be on the auto-impl)

**Review by Pratyush — APPROVED** [review]
Sure, this looks good!

**Comment by jon-chuang** [comment]
I will post a PR to retract these unecessary changes, I believe only the changes to formula made improvements, but defining unnecessary methods did not.

**Comment by ValarDragon** [comment]
DId you use the companion addition chains in https://github.com/arkworks-rs/curves/pull/43 ?

**Comment by jon-chuang** [comment]
well ok my mistake, I get why this might makes sense. However, I don't think a speedup has not been demonstrated for the cases where the non-residue is not -1.

Also, I'm not convinced the benchmarks are accurate since one is not using the n_fold feature. So I'd like to collect more data. 

Due to the ugly proliferation of the methods, I'm convinced that an automatic framework as per https://github.com/arkworks-rs/algebra/issues/206 would be neater by introducing less burden to the end user.

**Pull request merged**
